### DISCUSSION TITLE
How to login with IdP which belongs to an organization?

### DISCUSSION THREAD

**Discussion by TakumiHaruta** [zitadel/zitadel#5088] [Q&A] [answered]
Hi there,

After I registered an external IdP in an "Organization", how can I enable this SSO on the login page?

<img width="1503" alt="Screenshot 2023-01-23 at 9 18 57" src="https://user-images.githubusercontent.com/35373297/213948462-68d3e795-83d0-4b26-a147-8f74e20ebc21.png">

I understand the login button is displayed if IdP is enabled in "Instance" settings, but I don't know how the button is displayed if IdP belongs to only an Organization.

<img width="1512" alt="Screenshot 2023-01-23 at 9 24 42" src="https://user-images.githubusercontent.com/35373297/213949046-e5c701ca-5341-45d3-b15a-4d095e8cc4ce.png">

I thought I need to use the login page of an Organization, but the primary domain seems not to be able to be used for the login page.

<img width="1504" alt="Screenshot 2023-01-23 at 9 35 58" src="https://user-images.githubusercontent.com/35373297/213949190-01a291bc-9663-4daa-b629-b242b412d946.png">

Here are our context and use cases;

- We're hosting B2B SaaS, and tenants are separated by an "Organization" in Zitadel
- Each organization wants to register its own IdP such as Azure AD or Google workspace to use SSO for our SaaS
- I'm not sure which is possible in Zitadel, but one of them below is ideal for our SaaS;
    1. User accesses the shared domain `https://auth.sample.com/` hosted by Zitadel, and it redirects to `https://app.sample.com/` after login succeeded
    2. User accesses the specific domain `https://tenant1.auth.sample.com/` hosted by Zitadel, and it redirects to `https://app.sample.com/` after login succeeded

We'd be glad if someone could help us!
Thanks

**Comment by hifabienne** [comment, marked as answer]
Hi @TakumiHaruta 
To be able to login to your organization you have to send the organization scope in your authorization request.
This could look something like this:
{your-domain}/oauth/v2/authorize?client_id={your-clientid}&redirect_uri=http%3A%2F%2Flocalhost%3A8080%2Fapi%2Fauth%2Fcallback%2Fzitadel&response_type=code&scope=openid%20email%20profile%20**urn%3Azitadel%3Aiam%3Aorg%3Aid%3A168811945419506433**&code_challenge=9az09PjcfuENS7oDK7jUd2xAWRb-B3N7Sr3kDoWECOY&code_challenge_method=S256

You can also test the behaviour of the login with our authentication playground, make sure to tick the box for the organization scope: https://zitadel.com/docs/apis/openidoauth/authrequest

**Reply by TakumiHaruta** [reply]
Thank you @hifabienne 

I understood how to do it.
So you mean the client has to know the organization id before requesting `https://{your-domain}/oauth/v2/authorize`, which requires the client to dynamically change it in the request right?

How can we fetch the organization id by what (e.g. email address?) from the client, and do we have to build own UI to control routes to each organization?

**Reply by hifabienne** [reply]
To be able to show the user the organization settings we need to know somehow which organization it is.
You have different possibilities.
1. You need to know which organization and you can send the organization scope
2. the user types the username first and then the settings are shown. If you have only the possibility for an external idp (e.g Google), we are able to redirect automatically: read the following docs about how to configure in that case: https://zitadel.com/docs/guides/solution-scenarios/configurations#automatically-redirect-users-if-the-organization-has-only-one-identity-provider
If user do not always already exist, you can enable domain discovery in the instance settings and add the domains to the organization. In that case we can also redirect if the user does not exist. 
E.g. if I added zitadle.com and zitadel.ch to my organization domains, enter fabienne@zitadel.com for the login and the user doesnt exist, we recognise the domain and can show the settings
<img width="702" alt="image" src="https://user-images.githubusercontent.com/38692350/213981239-44903496-4d5c-4070-ab1b-e22749cda8d4.png">

**Reply by TakumiHaruta** [reply]
@hifabienne 

We can't choose option 2 because our users would register the same email address in multiple organizations.
Then we'll try to find better ways to implement option 1 not to use the login console without the organization scope.

I have another further question regarding the login console.
Is it possible to deny users access to the ZITADEL administration console?
We'd like to allow only the initial user `Zitadel Admin` to access the console at least.

**Reply by hifabienne** [reply]
Only users with a specific membership in ZITADEL will be able to access the Administration part of the console. Users without a role will only see their own profile to manage their own user data.

At the moment it is not possible to restrict the access for people with no manager roles.

**Reply by TakumiHaruta** [reply]
I see. We have two future requests then;

1. Want to have the option to deny users access to their own user account in Zitadel administration console. If you manage tenants by organization in Zitadel, the organization scope is required to let users access their own tenant, but this logic depends on the redirection from an application. If users bookmark Zitadel domain `https://{your-domain}` to access our service, they always access to Zitadel console unexpectedly.
2. Want to use a sub-domain to access the login page owned by an organization without redirecting with the organization scope. Currently I couldn't find the sub domain settings except `ExternalDomain` in [defaults.yml](https://github.com/zitadel/zitadel/blob/v2.18.0/cmd/defaults.yaml), so Zitadel instance can't accept the request from its sub-domain. If Zitadel accepts the request from sub-domain of `ExternalDomain` (which could be the primary domain of organitzation) and automatically link it to the organization's info, it would be really nice for users because they can bookmark its domain. I expect the auth system of Slack for example.

**Reply by hifabienne** [reply]
What might help you for the first problem for the moment is to change the default redirect url of your instance to your application.If we do not have a authorization request (e.g direct call of domain or from a password reset email) we always call the default redirect url.
If you change it to your application, the users will not land on the console, but on your application. 
You can find it in the login settings:
<img width="702" alt="image" src="https://user-images.githubusercontent.com/38692350/214211263-47c4fef8-44bf-44d5-bae9-b81ebccfba43.png">

I will create a separat issue for the request with additional domains linking to an organization: https://github.com/zitadel/zitadel/issues/5094

**Reply by TakumiHaruta** [reply]
Thank you for converting it into an issue!

> What might help you for the first problem for the moment is to change the default redirect url of your instance to your application.If we do not have a authorization request (e.g direct call of domain or from a password reset email) we always call the default redirect url.

This could be helpful not to let users access the console, but it's not perfect because our users would register the same email address in multiple organizations. Some users who have just one account could authenticate, but some who have more than two accounts could fail to authenticate. It's not so critical, but a bit weird for users in terms of UX.

But we use the default redirect url for now. Thanks!
